File: server/src/cve/kev-importer.ts
import axios from "axios";
import { db, dbAll, dbGet } from "../db/database";
import { KevSyncState, KevYearStat } from "../types";
import { checkEnrichmentAlerts } from "./enrichment-alerter";

const KEV_URL =
  "https://www.cisa.gov/sites/default/files/feeds/known_exploited_vulnerabilities.json";

interface KevVulnerability {
  cveID: string;
  vendorProject: string;
  product: string;
  vulnerabilityName: string;
  dateAdded: string;
  shortDescription: string;
  requiredAction: string;
  dueDate: string;
  knownRansomwareCampaignUse: string;
  notes: string;
}

interface KevFeed {
  catalogVersion: string;
  dateReleased: string;
  count: number;
  vulnerabilities: KevVulnerability[];
}

interface KevSyncResult {
  count: number;
  catalogVersion: string;
  dateReleased: string;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
let _upsert: any = null;

function ensureStmt(): void {
  if (_upsert) return;
  _upsert = db.prepare(
    `INSERT OR REPLACE INTO cisa_kev
       (cve_id, vendor_project, product, vulnerability_name, date_added,
        short_description, required_action, due_date, known_ransomware_campaign_use,
        notes, synced_at)
     VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`
  );
}

export async function syncKev(): Promise<KevSyncResult> {
  const response = await axios.get<KevFeed>(KEV_URL, {
    responseType: "json",
    headers: { "User-Agent": "monitor-app/1.0" },
    timeout: 30000,
  });

  const feed = response.data;
  const vulns = feed.vulnerabilities ?? [];
  const syncedAt = new Date().toISOString();

  ensureStmt();
  db.exec("BEGIN IMMEDIATE");
  try {
    for (const v of vulns) {
      _upsert.run(
        v.cveID,
        v.vendorProject ?? null,
        v.product ?? null,
        v.vulnerabilityName ?? null,
        v.dateAdded ?? null,
        v.shortDescription ?? null,
        v.requiredAction ?? null,
        v.dueDate ?? null,
        v.knownRansomwareCampaignUse ?? null,
        v.notes ?? null,
        syncedAt
      );
    }
    db.exec("COMMIT");
  } catch (err) {
    db.exec("ROLLBACK");
    throw err;
  }

  await checkEnrichmentAlerts();

  return {
    count: vulns.length,
    catalogVersion: feed.catalogVersion ?? "",
    dateReleased: feed.dateReleased ?? "",
  };
}

export function getKevSyncState(): KevSyncState {
  const row = dbGet<{ total: number; last_synced_at: string | null }>(
    "SELECT COUNT(*) AS total, MAX(synced_at) AS last_synced_at FROM cisa_kev"
  );
  const year_stats = dbAll<KevYearStat>(
    `SELECT strftime('%Y', date_added) AS year,
            COUNT(*) AS count,
            SUM(CASE WHEN known_ransomware_campaign_use = 'Known' THEN 1 ELSE 0 END) AS ransomware_count
     FROM cisa_kev
     WHERE date_added IS NOT NULL
     GROUP BY year
     ORDER BY year DESC`
  );
  return {
    total: row?.total ?? 0,
    last_synced_at: row?.last_synced_at ?? null,
    is_syncing: false, // caller sets this in-memory
    year_stats,
  };
}

File: vigil/src/cve/kev_importer.rs
use serde::Deserialize;

use crate::db::DbPool;

const KEV_URL: &str =
    "https://www.cisa.gov/sites/default/files/feeds/known_exploited_vulnerabilities.json";

#[derive(Deserialize)]
struct KevFeed {
    #[serde(rename = "catalogVersion")]
    catalog_version: Option<String>,
    #[serde(rename = "dateReleased")]
    date_released: Option<String>,
    vulnerabilities: Vec<KevVuln>,
}

#[derive(Deserialize)]
struct KevVuln {
    #[serde(rename = "cveID")]
    cve_id: String,
    #[serde(rename = "vendorProject")]
    vendor_project: Option<String>,
    product: Option<String>,
    #[serde(rename = "vulnerabilityName")]
    vulnerability_name: Option<String>,
    #[serde(rename = "dateAdded")]
    date_added: Option<String>,
    #[serde(rename = "shortDescription")]
    short_description: Option<String>,
    #[serde(rename = "requiredAction")]
    required_action: Option<String>,
    #[serde(rename = "dueDate")]
    due_date: Option<String>,
    #[serde(rename = "knownRansomwareCampaignUse")]
    known_ransomware_campaign_use: Option<String>,
    notes: Option<String>,
}

pub struct KevSyncResult {
    pub count: usize,
    pub catalog_version: String,
    pub date_released: String,
}

pub async fn sync_kev(db: &DbPool) -> anyhow::Result<KevSyncResult> {
    tracing::info!("Downloading CISA KEV feed");

    let feed: KevFeed = reqwest::Client::new()
        .get(KEV_URL)
        .header("User-Agent", "monitor-app/1.0")
        .timeout(std::time::Duration::from_secs(30))
        .send()
        .await?
        .error_for_status()?
        .json()
        .await?;

    let vulns = feed.vulnerabilities;
    let catalog_version = feed.catalog_version.unwrap_or_default();
    let date_released = feed.date_released.unwrap_or_default();
    let count = vulns.len();
    let synced_at = chrono::Utc::now().to_rfc3339();

    let db = db.clone();
    tokio::task::spawn_blocking(move || -> anyhow::Result<()> {
        let conn = db.lock().unwrap();
        conn.execute_batch("BEGIN IMMEDIATE")?;
        let result: rusqlite::Result<()> = (|| {
            let mut stmt = conn.prepare_cached(
                "INSERT OR REPLACE INTO cisa_kev
                 (cve_id, vendor_project, product, vulnerability_name, date_added,
                  short_description, required_action, due_date,
                  known_ransomware_campaign_use, notes, synced_at)
                 VALUES (?1,?2,?3,?4,?5,?6,?7,?8,?9,?10,?11)",
            )?;
            for v in &vulns {
                stmt.execute(rusqlite::params![
                    v.cve_id,
                    v.vendor_project,
                    v.product,
                    v.vulnerability_name,
                    v.date_added,
                    v.short_description,
                    v.required_action,
                    v.due_date,
                    v.known_ransomware_campaign_use,
                    v.notes,
                    synced_at,
                ])?;
            }
            Ok(())
        })();
        match result {
            Ok(()) => conn.execute_batch("COMMIT")?,
            Err(e) => {
                conn.execute_batch("ROLLBACK")?;
                return Err(e.into());
            }
        }
        Ok(())
    })
    .await??;

    tracing::info!(count, "CISA KEV sync complete");
    Ok(KevSyncResult { count, catalog_version, date_released })
}
